### PR TITLE
Unit Test worker.go

### PR DESCRIPTION
**Pull request by cmicat**
Signed-off-by: raymondmiaochaoyue <raymondmiaochaoyue@didiglobal.com>

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Add Unit Test for worker.go

**Which issue(s) this PR fixes**:

Part of https://github.com/karmada-io/karmada/issues/2579

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/util/worker_test.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+	stopChan := make(chan struct{})
+	worker.Run(5, stopChan)
+
+	for i := 0; i < cnt; i++ {
+		worker.Add(i)
+	}
+
+	time.Sleep(time.Second)
+
+	close(stopChan)
```
Comment (jwcesign) [inline pkg/util/worker_test.go:127]: defer close(stopChan)

File: pkg/util/worker_test.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	processed := reconcile.getProcessed()
+	if len(processed) < cnt {
+		t.Errorf("processed item not equal to input, len() = %v, processed item is %v", len(processed), processed)
+		return
+	}
+
+	sort.Ints(processed)
+
+	for i := 0; i < cnt; i++ {
+		if processed[i] != i {
```
Comment (jwcesign) [inline pkg/util/worker_test.go:138]: Map is better?  use array is little tricky

**Comment by jwcesign** [comment]
other is LGTM

File: pkg/util/worker_test.go (hunk 1 above)
Comment (cmicat) [reply pkg/util/worker_test.go:127]: I intend to stop the worker after 1s. defer will close it on method exit.

File: pkg/util/worker_test.go (hunk 2 above)
Comment (cmicat) [reply pkg/util/worker_test.go:138]: OK. I can reuse `asyncWorkerReconciler.shouldPass` map.

File: pkg/util/worker_test.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	reconcile := newAsyncWorkerReconciler()
+	worker := newTestAsyncWorker(reconcile.ReconcileFunc)
+	stopChan := make(chan struct{})
+	worker.Run(5, stopChan)
+
+	for i := 0; i < cnt; i++ {
+		worker.Add(i)
+	}
+
+	time.Sleep(time.Second)
```
Comment (RainbowMango) [inline pkg/util/worker_test.go:124]: We can wait for `len(processed) == cnt` for a period of time, we don't know if 1s is enough or not, you know the test will be launched on various environments.

File: pkg/util/worker_test.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+	close(stopChan)
+
+	processed := reconcile.ProcessedItem()
+	if len(processed) < cnt {
+		t.Errorf("processed item not equal to input, len() = %v, processed item is %v", len(processed), processed)
```
Comment (chaunceyjiang) [inline pkg/util/worker_test.go:130]:
``` diff
        if len(processed) < cnt {
-               t.Errorf("processed item not equal to input, len() = %v, processed item is %v", len(processed), processed)
-               return
+               t.Fatalf("processed item not equal to input, len() = %v, processed item is %v", len(processed), processed)
        }

```

File: pkg/util/worker_test.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+		return
+	}
+
+	for i := 0; i < cnt; i++ {
+		if _, ok := processed[i]; !ok {
+			t.Errorf("expected item not processed, expected: %v, all processed item: %v", i, processed)
```
Comment (chaunceyjiang) [inline pkg/util/worker_test.go:136]: ditto.

File: pkg/util/worker_test.go (hunk 3 above)
Comment (cmicat) [reply pkg/util/worker_test.go:124]: I already set ratelimiter QPS & delay, so ratelimiter will not delay the execution. I think 1s is enough for most such short code to finish.

Comment (cmicat) [reply pkg/util/worker_test.go:124]: @RainbowMango

Comment (cmicat) [reply pkg/util/worker_test.go:124]: I think I can write a polling logic: check the status periodically, fail after certain seconds. For example, check every 50 ms, fail after 10s. This way, bigger wait duration won't result in bigger test execution time.

Comment (cmicat) [reply pkg/util/worker_test.go]: > I think I can write a polling logic: check the status periodically, fail after certain seconds. For example, check every 50 ms, fail after 10s. This way, bigger wait duration won't result in bigger test execution time.

Done

**Comment by cmicat** [comment]
@RainbowMango Please take a look again

**Comment by cmicat** [comment]
@jwcesign @chaunceyjiang 
The problems you mentioned is all fixed.

**Review by RainbowMango — APPROVED** [review]
/approve

**Pull request merged**